design/census-import/generate-census-records.js

```diff
@@ -26,53 +26,78 @@ function writeCsv (data) {
   const headers = ['LEA', 'Estab', 'UPN', 'Surname', 'Forename', 'Middlenames', 'Gender', 'DOB']
   const ws = fs.createWriteStream(outputFilename, { flags: 'a' })
   csv
-    .write(data, {headers: headers})
+    .write(data, { headers: headers })
     .pipe(ws)
 }
 
-async function getSchoolsLength () {
-  const sql = `SELECT COUNT(*) AS [cnt]
+async function getSchoolDfeNumbers () {
+  const sql = `SELECT dfeNumber
   FROM ${sqlService.adminSchema}.school`
   const result = await sqlService.query(sql, [])
-  const obj = R.head(result)
-  return R.prop('cnt', obj)
+  return R.map(r => R.prop('dfeNumber', r), result)
+}
+
+async function getSchoolDfeNumberCount () {
+  const sql = `SELECT COUNT(dfeNumber) AS cnt
+  FROM ${sqlService.adminSchema}.school`
+  const result = await sqlService.query(sql, [])
+  return R.prop('cnt', R.head(result))
+}
+
+function generateUpn (dfeNumber) {
+  const leaCode = dfeNumber.slice(0, 3)
+  const estabCode = dfeNumber.slice(3, dfeNumber.length)
+  const minYearValue = academicYear - 11
+  const maxYearValue = academicYear - 7
+  const year = Math.floor(Math.random() * (maxYearValue - minYearValue + 1) + minYearValue).toString().substr(-2)
+  let baseUpn = leaCode + estabCode + year
+  pupilCount += 1
+  if (pupilCount > 999) {
+    pupilCount = 1
+    baseUpn = (parseInt(baseUpn) + 1000).toString()
+  }
+  const serial = pupilCount.toString().padStart(3, '0')
+  return upnService.calculateCheckLetter(baseUpn + serial) + baseUpn + serial
 }
 
-async function main (options) {
-  winston.info('Generating census records...')
-  const { recordsLength } = options
+async function main () {
+  await sqlService.initPool()
+  console.log('Generating census records...')
   const csvData = []
-  let baseUpn = '702500001'
-  let schoolsLength
+  let dfeNumbers
+  let dfeNumberCount
+  let upns = []
   try {
-    schoolsLength = await getSchoolsLength()
-    schoolsLength = schoolsLength - 1
+    dfeNumbers = await getSchoolDfeNumbers()
+    dfeNumberCount = await getSchoolDfeNumberCount()
+    dfeNumberCount = dfeNumberCount - 1
   } catch (error) {
-    winston.info(error)
+    console.log(error)
```

Comment:
should you exit at this point or go into the loop?

Reply:
@GuyHarwood updated this to call drainPool() on catch and therefore exit.

Reply:
does drainPool call `process.exit` ? surely you need to explicitly exit the routine?

Reply:
when drainPool is called the node process will exit

Reply:
@GuyHarwood I have updated this block by assigning the necessary failure exit code since this is all it really needs.
`process.exit(1)` is not required and it is considered a bad practice for that case as the script will terminate gracefully when there is no additional work pending in the event loop.

Reply:
but we do need to exit the routine.  does setting `process.exitCode` make the current procedure return early?  i would have thought not 🤔  there is a loop directly below the catch block, which as far as i can tell will execute after you have caught the error.  surely you do not want that loop to run?

Reply:
@GuyHarwood yea I overlooked it. We do need to exit immediately actually. You are correct.
I updated to use `process.exit(1)` instead

Reply:
no worries.  many thanks for the changes. will approve now